Add nested field support to KQL (#47070)
elastic/kibana: docs/discover/kuery.asciidoc

in the response field, but a query for just `200` will search for 200 across all fields in your index.
============

===== Nested Field Support

> KQL supports querying on {ref}/nested.html[nested fields] through a special syntax. Nested fields can be queried in subtly different

Reviewer: ```suggestion
KQL supports querying on {ref}/nested.html[nested fields] through a special syntax. You can query nested fields in subtly different
```

> ways depending on the results you would like to see, so crafting nested queries requires some extra thought.

Reviewer: ```suggestion
ways, depending on the results you want, so crafting nested queries requires extra thought.
```

> One of the main things to consider is how you want parts of the nested query to match the individual nested documents.

Reviewer: ```suggestion
One main consideration is how to match parts of the nested query to the individual nested documents.
```

There are two main approaches to take:

> 1. Parts of the query may only match a single nested document. This is what most people want when querying on a nested field.

Reviewer: ```suggestion
* *Parts of the query may only match a single nested document.* This is what most users want when querying on a nested field.
```

> 2. Parts of the query may match different nested documents. This is how a regular object field works but nested fields 

Reviewer: ```suggestion
*  *Parts of the query match different nested documents.* This is how a regular object field works. 
```

Author: Removing the "may" makes it sound to me like parts of the query *must* match different documents. Do you think it's fair to keep it in?

Reviewer: How about "can"?  Otherwise, keep may" in.

Author: "can" sounds good, I'll do that 👍

> can be queried in the same way. Although generally less useful, there are occasions where one might want to query a nested field in this way.

Reviewer: ```suggestion
 Although generally less useful, there might be occasions where you want to query a nested field in this way.
```

> The KQL syntax supports both approaches. Say we have the following document, where `items` is a nested field:

Reviewer: ```suggestion
Let's take a look at the first approach. In the following document, `items` is a nested field:
```

[source,json]
----------------------------------
{
  "grocery_name": "Elastic Eats",
  "items": [
    {
      "name": "banana",
      "stock": "12",
      "category": "fruit"
    },
    {
      "name": "peach",
      "stock": "10",
      "category": "fruit"
    },
    {
      "name": "carrot",
      "stock": "9",
      "category": "vegetable"
    },
    {
      "name": "broccoli",
      "stock": "5",
      "category": "vegetable"
    }
  ]
}
----------------------------------

> If we wanted to find stores that have more than 10 bananas in stock, we would write a query like this:

Reviewer: ```suggestion
To find stores that have more than 10 bananas in stock, you would write a query like this:
```

`items:{ name:banana and stock > 10 }`

`items` is the "nested path". Everything inside the curly braces (the "nested group") must match a single document. 

> So for example, `items:{ name:banana and stock:9 }` would not match because there isn't a single nested document that 

Reviewer: ```suggestion
For example, `items:{ name:banana and stock:9 }` does not match because there isn't a single nested document that 
```

matches the entire query in the nested group. 

> What if you want to find a store with more than 10 bananas that *also* stocks vegetables? This is the second way of querying a nested field, and you can do it like this:

Reviewer: @Bargs I agree with what Tim is saying.  Happy to review the updated text when it is ready.

> matches the entire query in the nested group. This is the first way to query a nested field listed above.

Reviewer: ```suggestion
matches the entire query in the nested group. 
```

> What if we wanted to find a store with more than 10 bananas that *also* stocks vegetables? This is the second way of querying a nested field and we can do it like this:

Reviewer: ```suggestion
What if you want to find a store with more than 10 bananas that *also* stocks vegetables? This is the second way of querying a nested field, and you can do it like this:
```

`items:{ name:banana and stock > 10 } and items:{ category:vegetable }`

The first nested group (`name:banana and stock > 10`) must still match a single document, but the `category:vegetables`

> sub-query may match a different nested document since it is in a separate group.

Reviewer: ```suggestion
subquery can match a different nested document because it is in a separate group.
```

> KQL's syntax also supports nested fields inside of other nested fields. You simply have to specify the full path. Say we

Reviewer: ```suggestion
KQL's syntax also supports nested fields inside of other nested fields&mdash;you simply have to specify the full path. Suppose you
```

> have a document like the following where `level1` and `level2` are both nested fields.

Reviewer: ```suggestion
have a document where `level1` and `level2` are both nested fields:
```

[source,json]
----------------------------------
{
  "level1": [
    {
      "level2": [
        {
          "prop1": "foo",
          "prop2": "bar"
        },
        {
          "prop1": "baz",
          "prop2": "qux"
        }
      ]
    }
  ]
}
----------------------------------

> We can match on a single nested document by specifying the full path like this:

Reviewer: ```suggestion
You can match on a single nested document by specifying the full path:
```
